netbox/ipam/forms/bulk_import.py

```diff
@@ -572,15 +570,43 @@ class ServiceImportForm(NetBoxModelImportForm):
     class Meta:
         model = Service
         fields = (
-            'device', 'virtual_machine', 'ipaddresses', 'name', 'protocol', 'ports', 'description', 'comments', 'tags',
+            'parent_object_type', 'ipaddresses', 'name', 'protocol', 'ports', 'description', 'comments',
+            'tags', 'parent_object_id',
         )
 
-    def clean_ipaddresses(self):
-        parent = self.cleaned_data.get('device') or self.cleaned_data.get('virtual_machine')
-        for ip_address in self.cleaned_data['ipaddresses']:
+    def __init__(self, data=None, *args, **kwargs):
+        super().__init__(data, *args, **kwargs)
+
+        # Limit parent queryset by assigned parent object type
+        if data:
+            match data.get('parent_object_type'):
+                case 'dcim.device':
+                    self.fields['parent'].queryset = Device.objects.all()
+                case 'ipam.fhrpgroup':
+                    self.fields['parent'].queryset = FHRPGroup.objects.all()
+                case 'virtualization.virtualmachine':
+                    self.fields['parent'].queryset = VirtualMachine.objects.all()
+
+    def save(self, *args, **kwargs):
+        if (parent := self.cleaned_data.get('parent')):
+            self.instance.parent = parent
+
+        return super().save(*args, **kwargs)
+
+    def clean(self):
+        super().clean()
+
+        if (parent := self.cleaned_data.get('parent')):
+            self.cleaned_data['parent_object_id'] = parent.pk
+        elif not parent and (parent_id := self.cleaned_data.get('parent_object_id')):
+            ct = self.cleaned_data.get('parent_object_type')
+            parent = ct.model_class().objects.filter(id=parent_id).first()
+            self.cleaned_data['parent'] = parent
+
+        for ip_address in self.cleaned_data.get('ipaddresses', []):
             if not ip_address.assigned_object or getattr(ip_address.assigned_object, 'parent_object') != parent:
                 raise forms.ValidationError(
                     _("{ip} is not assigned to this device/VM.").format(ip=ip_address)
                 )
 
-        return self.cleaned_data['ipaddresses']
+        return self.cleaned_data
```

Comment:
All of this logic got moved up to `clean()` from `clean_ipaddresses()` because it needs access to a resolved value for the `parent` field. I tried doing that within the existing `clean_ipaddresses()`, but it was a mess and didn't feel reliable.